2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+---
+RFC: RFCxxxx
+Author: Steve Lee
+Status: Draft
+SupercededBy: N/A
+Version: 0.1
+Area: PowerShellGet
+Comments Due: 7/30
+Plan to implement: Yes
+---
+
+# PowerShellGet 3.0 Module
+
+PowerShellGet has enabled an active community to publish and install PowerShell
+Modules.
+However, it has some architectural and design limitations that prevent it from
+moving forward more quickly.
+This RFC proposes some significant changes to address this.
+
+> [!NOTE]
+> There are no intentions to break existing PSGet v2 users so existing
+> users can stay on PSGet v2 and that will continue to work.
+
+## Motivation
+
+    As a PowerShell user,
+    I can discover how to install missing cmdlets,
+    so that I can be more productive.
+
+    As a PowerShellGet contributor,
+    I can more easily contribute code,
+    so that PowerShellGet evolves more quickly.
+
+## User Experience
+
+```powershell
+PS> Get-Satisfaction
+Get-Satisfaction : The term 'Get-Satisfaction' is not recognized as the name of a cmdlet, function, script file, or operable program.
+Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
+At line:1 char:1
++ Get-Satisfaction
++ ~~~
++ CategoryInfo          : ObjectNotFound: (Get-Satisfaction:String) [], CommandNotFoundException
++ FullyQualifiedErrorId : CommandNotFoundException
+
+Suggestion: You can obtain `Get-Satisfaction` by running `Install-PSResource HappyDays`.
+
+PS> Install-PSResource HappyDays
+```
+
+## Specification
+
+### Rewrite of PowerShellGet
+
+PowerShellGet is currently written as PowerShell script with a dependency on PackageManagement.
+Proposal is to write PSResource in C# to reduce complexity and make easier to maintain.
+In addition, remove dependency on PackageManagement completely as well as dependency on
+nuget.exe.
+This module will only use REST APIs to get and publish nupkgs.
+This module would be shipped in PowerShell 7.
+
+### Side-by-side with PowerShellGet
+
+To ease transition due to the enormity of the breaking change, this module will
+have a different names for the cmdlets to allow for side by side use
+with the existing PowerShellGet module.
+Since the current PowerShellGet 2.x version is a script module and this new one
+is a C# based module, they can coexist side-by-side.
+
+### Cmdlet compatibility
+
+Function wrappers will be provided to provide compatibility with the two most often
+used cmdlets: `Install-Module` and `Find-Module`.
+These cmdlets will output a warning to use the new cmdlets.
+
+### Local cache
+
+Rather than connecting to PSGallery and other repositories for finding modules,
+the user must explicitly request the current metadata of
+modules from the repositories and that is cached locally.
+`Find-PSResource` (see below) works against this local cache.
+This will also enable changes in PowerShell to use `Find-PSResource -Type Command` to look
```

Comment:
Would be good to clarify the cache semantics, I don't quite see what they are. Is it:
- manually managed by the user (like apt-get) and never refers to network package list unless you run an explicit command
-we always check with the repo and check that the cache is up-to-date, so current semantics are preserved
- somewhere in between

Reply:
I feel it would also be reasonable to experiment with NOT caching and having a more performant web-service handle the search queries. We have ample existence proof that you can write type-ahead completion code that does round-trips to a web-service that works fast enough for interactive use. This could be as striaghtforward as integrating with Azure Search for the PS gallery.

Reply:
Regarding cache semantics, original thinking was to follow `apt` and have it be manual each time.  I've since leaned towards automatic updates.  Once we have an implementation we can measure perf.

Regarding going off box, look at your arguments against it here: https://github.com/PowerShell/PowerShellGet/issues/287

Reply:
It's **especially** important not to go off-box when doing suggestions in command-not-found, or you'll get a ton of backlash from those of us concerned with privacy|security 🏴‍☠️

Reply:
How do you envision the cache dealing with multiple possible repositories? Including those that do pass through to PowerShellGallery or other feeds that lead to "duplicate" modules being available? I'd imagine a single cache file that has a repositories attribute listing which ones it's available from would make the most sense and keep size down.

Reply:
Good suggestion, will add to RFC.